### PR TITLE
docker_swarm inventory plugin - new attribute with parsed node URI

### PR DESCRIPTION
**Pull request by WojciechowskiPiotr**
##### SUMMARY
New attribute for each inventory entry containing the parsed node URI ready to use as value of `docker_host` option in docker swarm modules

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_swarm

##### ADDITIONAL INFORMATION
If the Ansible module for Docker Swarm family uses API to connect to docker host instead of SSH and local socket on that host, it is required to provide the URI of the docker host module has to connect to as `docker_host` option. The `ansible_host` attribute assigned to each inventory entry contains only the IP address.

To make the playbook developers life a little easier I propose to add new attribute `ansible_docker_uri` with full URI that can easily be provided in playbooks. As a result, it won't be required for the playbook developers to modify the `ansible_host` value in many cases.

Example of such usage below. The playbook runs the `docker_node_facts` locally and using the `leader` group connects to the swarm leader to read the facts about the node. Another use case might be the playbook performing an operation on every node with the role of the `manager`.

```
- name: Inspect leader node
  hosts: leader
  connection: local
  gather_facts: no

    - name: Get leader node facts
      docker_node_facts:
        self: yes
        docker_host: "{{ ansible_host_uri }}"
        tls: true

```

NOTE: The returned value is in the form of `tcp://<IP Address>:2376`. The port is static now but it may be provided as a parameter or set to `2376` or `2375` depending if TLS is in use running the inventory plugin

NOTE: It covers #53893 so needs to be rebased later

### CODE REVIEW

**Comment by WojciechowskiPiotr** [comment]
@morph027

**Comment by morph027** [comment]
Instead of defaulting the port to 2376, it might be better to split the `host` variable into 2 variables `host` and optional `port` to be able to re-use `port` in line 189 and 197.

**Comment by WojciechowskiPiotr** [comment]
> Instead of defaulting the port to 2376, it might be better to split the `host` variable into 2 variables `host` and optional `port` to be able to re-use `port` in line 189 and 197.

Sure, just waiting for initial feedback about the proposed feature. It can be port specified in `host` unless `port` as additional parameter is provided

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [inline lib/ansible/plugins/inventory/docker_swarm.py]: The name `host_uri` suggests that this option *expects* a host URI, not that it provides one. How about:
```suggestion
        add_host_uri:
```
Or `insert_host_uri`, `inject_host_uri`, `provide_host_uri`, ...?

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [inline lib/ansible/plugins/inventory/docker_swarm.py]: Doesn't `self.node_attrs['ManagerStatus']['Addr']` actually contain the correct port (IIRC, don't want to create a swarm to check :) )? If that's the case, why not use it here? Also, why not call `parse_address(self.node_attrs['ManagerStatus']['Addr'])` simply once?

Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py]: It contains the `2377` which is the communication port for swarm nodes port, not the management port.
We can optimize it to just using the `parse_address()` once but it won't give any benefit to a number of code lines, just one less method execution and some memory reserved for a variable. I see no real benefit.

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py]: I don't mind any name :) @morph027 any preference?

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [reply lib/ansible/plugins/inventory/docker_swarm.py]: Ah, makes sense. I just remembered that there was a port, and somehow assumed it's the one you need here...

Just executing code once is already a real benefit. Saves some cycles. Also, it improves readability since a fitting name will show the reader directly what `parse_address(...)[0]` is :)

Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py]: Ok, will add this with the port detection change as well

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [inline lib/ansible/plugins/inventory/docker_swarm.py]: Shouldn't this be an `int` (and converted to a string when needed)?

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [inline lib/ansible/plugins/inventory/docker_swarm.py]: Why not `... or self.node_attrs['Status']['Addr']`? It can't be worse than `0.0.0.0` :)

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [inline lib/ansible/plugins/inventory/docker_swarm.py:216]: Shouldn't the url begin with `https://` when TLS is enabled? At least that's what `module_utils/docker/common.py` is doing.

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py]: You never know ;)

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [reply lib/ansible/plugins/inventory/docker_swarm.py]: (TBH, I'm not sure whether inventory plugins actually support that...)

Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py]: I was wondering what is safer. After all, it will be string at the end. I am not sure of here the type reference is just for documentation or it really means something. I just merge it as any other string but was wandering on some additional checkings

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py:216]: The `docker.api.client` lib actually recommends `tcp://` as `base_url`. The library does the proper changes so I prefer providing the `tcp://` and let it take care of the rest

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [reply lib/ansible/plugins/inventory/docker_swarm.py]: I think it is documentation only. See https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/__init__.py#L55-L62 (I've been checking it out for #53906.)

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (felixfontein) [reply lib/ansible/plugins/inventory/docker_swarm.py:216]: Good point. TBH, I have no idea why module_utils/docker/common.py does this replacement ­— maybe it once was necessary, but isn't anymore. Or maybe it was never necessary...

File: lib/ansible/plugins/inventory/docker_swarm.py
Comment (WojciechowskiPiotr) [reply lib/ansible/plugins/inventory/docker_swarm.py]: I may change it to in the, I see no difference

Comment (felixfontein) [reply lib/ansible/plugins/inventory/docker_swarm.py]: At least then nobody can claim that they were allowed to put an arbitrary string in there ;)

**Comment by WojciechowskiPiotr** [comment]
I reformatted some longer lines to make code more clean and easier to read

**Review by felixfontein — APPROVED** [review]
Looks good to me. @morph027 do you have anything else?

**Comment by morph027** [comment]
Nope, lgtm too. Guess the changes in tls section will break https://github.com/ansible/ansible/pull/53906 without a rebase, but this one is :+1:

**Comment by felixfontein** [comment]
Yep, there will be conflicts, but I'll take care of that once this is merged ;-)

I'll make sure the bot runs CI again, so we know that this works against the new tests.

rebuild_merge

**Comment by felixfontein** [comment]
Ok, that wasn't very helpful. Anyway, it's merged now :)

**Comment by felixfontein** [comment]
@WojciechowskiPiotr thanks for this PR!
@morph027 thanks for reviewing!

I've added a minimal test for the new option in #53913.

**Comment by WojciechowskiPiotr** [comment]
You're welcome :)

**Pull request merged**
